src/kallsyms.c

```diff
@@ -165,8 +176,29 @@ static int ksyms_cache_open(ksyms_t *ks)
 	if (ks->cache_fd < 0)
 		return -errno;
 
-	ks->cache = mmap(NULL, st.st_size, PROT_READ, MAP_PRIVATE,
+	ks->cache = mmap(NULL, st.st_size, PROT_READ | PROT_WRITE, MAP_PRIVATE,
 			 ks->cache_fd, 0);
+
+	/* For bsearch() to work properly, our cache must be sorted by
+	 * start address.  kallsyms is not guaranteed to be in order from
+	 * low address to high; modules seem to be particularly problematic.
+	 * Question: do we need to sort prior to cache creation?  Might make
+	 * that code a bit uglier but implicit assumption of ordering is used
+	 * to figure out end addresses.
+	 */
+	if (ks->cache) {
```

Comment:
Maybe it would be better to do the sorting at cache creation time?

Reply:
Although I notice now that may actually require more code... up to the original author.

Reply:
right, that was my thinking. The other thing is that whether we're creating or reading an existing cache, we need to sort it, otherwise we'd need to require users to remove the old (unsorted) caches for things to work right. Thanks for taking a look!